### PR TITLE
LabelEncoder kernel creation improvement

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Description
This PR updates the initialisation of the `_map` in `LabelEncoder_2` to be more memory efficient.

1. Firstly, we switch from `std::unordered_map` to [`absl::flat_hash_map`](https://abseil.io/tips/136). The latter has a more compact layout and doesn't have the overhead of maintaining reference validity like `std::unordered_map`, which is a feature we do not need here since we only initialise `_map` once at creation and then perform lookups duing `compute`. Abseil is already used extensively within onnxruntime.
2. Secondly, space is reserved before inserting to prevent rehashing and reallocation.

### Motivation and Context
For very large lookups, the LabelEncoder's kernel creation can require more RAM than necessary. These simple changes in `_map` initialisation improve initialisation speed and memory allocated.

### CODE REVIEW

**Comment by adityagoel4512** [comment]
Closes https://github.com/microsoft/onnxruntime/pull/16516

**Comment by adityagoel4512** [comment]
Closes https://github.com/microsoft/onnxruntime/issues/16575

**Comment by adityagoel4512** [comment]
@baijumeswani not sure if you are the right person to ask, but would it be possible to get a review on this?

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 1]
```diff
@@ -98,7 +98,7 @@ class LabelEncoder_2 final : public OpKernel {
   // A collection of key-value pairs. Each (a_key, a_value) pair
   // means that the "a_key" in the input would be mapped to "a_value".
   // If _map doesn't contain "a_key", we use _default_value as its output.
-  std::unordered_map<TKey, TValue> _map;
+  absl::flat_hash_map<TKey, TValue> _map;
```
Comment (baijumeswani) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:101]: Would you please use `InlinedHashMap` instead?
https://github.com/microsoft/onnxruntime/blob/2e2e6aeff646178d5606f85c33ff4c1189dbd0ef/include/onnxruntime/core/common/inlined_containers.h#L63-L76

Comment (baijumeswani) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h:101]:
```suggestion
  InlinedHashMap<TKey, TValue> _map;
```

Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h:101]: Thanks for pointing out the header file.

**Comment by adityagoel4512** [comment]
@baijumeswani looks like the CI has passed

**Review by baijumeswani — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thank you for your contribution.

**Pull request merged**